// react-bundle/dist/react-bundle.es.js
import { Component } from 'react';
import isPromise from 'ispromise';

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();









var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};











var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var Bundle = function (_Component) {
  inherits(Bundle, _Component);

  function Bundle() {
    classCallCheck(this, Bundle);

    var _this = possibleConstructorReturn(this, (Bundle.__proto__ || Object.getPrototypeOf(Bundle)).call(this));

    _this.state = {
      mod: null
    };


    _this.__mounted = false;
    return _this;
  }

  createClass(Bundle, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.__mounted = true;
      this.load(this.props);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.__mounted = false;
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.load !== this.props.load) {
        this.load(nextProps);
      }
    }
  }, {
    key: 'load',
    value: function load(_ref) {
      var _this2 = this;

      var _load = _ref.load;

      if (!this.__mounted) {
        return;
      }

      this.setState({
        mod: null
      });

      var loaded = function loaded(mod) {
        if (!_this2.__mounted) {
          return;
        }

        _this2.setState({ mod: mod.default ? mod.default : mod });
      };

      var res = _load(loaded);

      if (isPromise(res)) {
        res.then(loaded);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return this.props.children(this.state.mod);
    }
  }]);
  return Bundle;
}(Component);

export default Bundle;
